### PR TITLE
Standardize Logging in Manager

### PR DESCRIPTION
**Pull request by simanadler**
Signed-off-by: Sima Nadler <sima@il.ibm.com>
Addresses first 3 tasks of #926

Updates logging to use zerolog and new fybrik golang logging assistance functions.
Globally unique identifier for FybrikApplication instance is passed to all components to be included in logs.

Note: This PR does NOT include:
* Updates to the logging in the connectors
* Existing log entries in the manager were updated, but additional ones mentioned in tasks 4 and 5 were not added.

### CODE REVIEW

File: charts/fybrik/values.yaml [hunk 1]
```diff
@@ -266,7 +272,18 @@ opaServer:
   #   log: |-
   #     package system.log
   #     mask["/input"]
-  bootstrapPolicies: {}
+
+  # Bootstrap policies to load upon startup
+  bootstrapPolicies:
+    notebookSamplePolicy: |-
+      package dataapi.authz
```
Comment (shlomitk1) [inline charts/fybrik/values.yaml:279]: Why is notebookSamplePolicy defined in values.yaml and not in notebook-tests.values.yaml?

File: manager/apis/app/v1alpha1/zz_generated.deepcopy.go [hunk 2]
```diff
@@ -1,3 +1,4 @@
+//go:build !ignore_autogenerated
```
Comment (shlomitk1) [inline manager/apis/app/v1alpha1/zz_generated.deepcopy.go:1]: Why is this line here?

File: charts/fybrik/values.yaml (hunk 1 above)
Comment (simanadler) [reply charts/fybrik/values.yaml:279]: My mistake.  I moved it there temporarily when I was checking something.  Will remove.

File: manager/controllers/app/blueprint_controller.go [hunk 3]
```diff
@@ -85,7 +89,7 @@ func (r *BlueprintReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 		}
 	}
 
-	log.Info("blueprint reconcile cycle completed", "result", result)
+	log.Trace().Msg("blueprint reconcile cycle completed.") // TODO - Add result to log?
```
Comment (shlomitk1) [inline manager/controllers/app/blueprint_controller.go:92]: It is a debug printout. Any way to indicate this?

File: manager/apis/app/v1alpha1/zz_generated.deepcopy.go (hunk 2 above)
Comment (simanadler) [reply manager/apis/app/v1alpha1/zz_generated.deepcopy.go:1]: No idea.  I removed it

File: manager/controllers/app/blueprint_controller.go [hunk 4]
```diff
@@ -144,16 +148,20 @@ func getDomainFromImageName(image string) (string, error) {
 	return distributionref.Domain(named), nil
 }
 
-func (r *BlueprintReconciler) applyChartResource(ctx context.Context, log logr.Logger, chartSpec app.ChartSpec, args map[string]interface{}, blueprint *app.Blueprint, releaseName string) (ctrl.Result, error) {
-	log.Info(fmt.Sprintf("--- Chart Ref ---\n\n%v\n\n", chartSpec.Name))
+func (r *BlueprintReconciler) applyChartResource(ctx context.Context, log zerolog.Logger, chartSpec app.ChartSpec, args map[string]interface{}, blueprint *app.Blueprint, releaseName string) (ctrl.Result, error) {
+	// Get the unique id for the specific fybrikapplication instance.  Used for logging.
+	uuid := blueprint.GetAnnotations()[utils.FYBRIKAPPUUID]
```
Comment (shlomitk1) [inline manager/controllers/app/blueprint_controller.go:153]: Sometimes a function from utils is used to extract UUID. Can this be made consistent over the code?

File: manager/controllers/app/blueprint_controller.go (hunk 3 above)
Comment (simanadler) [reply manager/controllers/app/blueprint_controller.go:92]: Would you prefer this have verbosity level debug rather than trace?

File: manager/controllers/app/blueprint_controller.go (hunk 4 above)
Comment (simanadler) [reply manager/controllers/app/blueprint_controller.go:153]: Yes, changing

File: manager/controllers/app/blueprint_controller.go [hunk 5]
```diff
@@ -144,16 +148,20 @@ func getDomainFromImageName(image string) (string, error) {
...
+	alog := log.With().Str(logging.CONTROLLER, "Blueprint").Str(logging.FYBRIKAPPUUID, uuid).Str("blueprint", blueprint.GetName()).Logger()
```
Comment (shlomitk1) [inline manager/controllers/app/blueprint_controller.go:154]: Why can't log be used instead of alog?

File: manager/controllers/app/blueprint_controller.go [hunk 6]
```diff
@@ -273,7 +281,10 @@ func (r *BlueprintReconciler) updateModuleState(blueprint *app.Blueprint, instan
 	blueprint.Status.ModulesState[instanceName] = state
 }
 
-func (r *BlueprintReconciler) reconcile(ctx context.Context, log logr.Logger, blueprint *app.Blueprint) (ctrl.Result, error) {
+func (r *BlueprintReconciler) reconcile(ctx context.Context, log zerolog.Logger, blueprint *app.Blueprint) (ctrl.Result, error) {
+	uuid := blueprint.GetAnnotations()[utils.FYBRIKAPPUUID] // used for logging
+	alog := log.With().Str(logging.CONTROLLER, "Blueprint").Str(logging.FYBRIKAPPUUID, uuid).Str("blueprint", blueprint.GetName()).Logger()
```
Comment (shlomitk1) [inline manager/controllers/app/blueprint_controller.go:286]: Same comment. log seems to already include this information. Not clear why alog is needed.

File: manager/controllers/app/blueprint_controller.go [hunk 7]
```diff
@@ -297,6 +308,8 @@ func (r *BlueprintReconciler) reconcile(ctx context.Context, log logr.Logger, bl
 		}
 		module.Arguments.Labels[app.BlueprintNameLabel] = blueprint.Name
 		module.Arguments.Labels[app.BlueprintNamespaceLabel] = blueprint.Namespace
+		module.Arguments.Labels[utils.FYBRIKAPPUUID] = uuid // used for log correlation
```
Comment (shlomitk1) [inline manager/controllers/app/blueprint_controller.go:311]: Wouldn't it be better to define all label constants in one place? Also, I believe the convention is not to use all capital letters, I got a warning about it once.

File: manager/controllers/app/blueprint_controller.go [hunk 8]
```diff
@@ -437,10 +451,10 @@ func (r *BlueprintReconciler) checkResourceStatus(res *unstructured.Unstructured
 		}
 	}
 	// use expected values to compute the status
-	if r.matchesCondition(res, expected.SuccessCondition) {
+	if r.matchesCondition(res, expected.SuccessCondition, uuid) {
```
Comment (shlomitk1) [inline manager/controllers/app/blueprint_controller.go:454]: It could be clearer to pass log instance instead of uuid

File: manager/controllers/app/blueprint_controller.go (hunk 3 above)
Comment (shlomitk1) [reply manager/controllers/app/blueprint_controller.go:92]: Yes, debug instead of trace.

File: manager/controllers/app/fybrikapplication_controller.go [hunk 9]
```diff
@@ -90,12 +100,12 @@ func (r *FybrikApplicationReconciler) Reconcile(ctx context.Context, req ctrl.Re
 	if os.Getenv("ENABLE_WEBHOOKS") != "true" && (string(applicationContext.Status.ValidApplication) == "" || observedStatus.ValidatedGeneration != appVersion) {
 		// do validation on applicationContext
 		err := applicationContext.ValidateFybrikApplication(ApplicationTaxonomy)
-		log.V(0).Info("Reconciler validating Fybrik application")
+		log.Trace().Msg("Reconciler validating Fybrik application")
```
Comment (shlomitk1) [inline manager/controllers/app/fybrikapplication_controller.go:103]: Either debug, or could be removed completely.

File: manager/controllers/app/fybrikapplication_controller.go [hunk 10]
```diff
@@ -189,7 +202,7 @@ func (r *FybrikApplicationReconciler) checkReadiness(applicationContext *api.Fyb
 			provisionedBucketRef, found := applicationContext.Status.ProvisionedStorage[assetID]
 			if !found {
 				message := "No copy has been created for the asset " + assetID + " required to be registered"
-				r.Log.V(0).Info(message)
+				log.Info().Str(logging.DATASETID, assetID).Msg("No copy has been created for the asset " + assetID + " required to be registered")
```
Comment (shlomitk1) [inline manager/controllers/app/fybrikapplication_controller.go:205]: The message is duplicated.

File: manager/controllers/app/fybrikapplication_controller.go [hunk 11]
```diff
@@ -320,26 +338,26 @@ func (r *FybrikApplicationReconciler) reconcile(applicationContext *api.FybrikAp
...
+		log.Fatal().Err(err).Bool(logging.FORUSER, true).Bool(logging.AUDIT, true).Str(logging.ACTION, logging.CREATE).Msg("Could not determine in which cluster the workload runs")
 		return ctrl.Result{}, err
 	}
 	var requirements []DataInfo
 	for _, dataset := range applicationContext.Spec.Data {
 		req := DataInfo{
 			Context: dataset.DeepCopy(),
 		}
-		r.Log.V(0).Info("Preparing requirements for " + req.Context.DataSetID)
+		r.Log.Trace().Str(logging.DATASETID, req.Context.DataSetID).Str(logging.ACTION, logging.CREATE).Msg("Preparing requirements for dataset")
```
Comment (shlomitk1) [inline manager/controllers/app/fybrikapplication_controller.go:357]: Can be removed.

File: manager/controllers/app/blueprint_controller.go (hunk 3 above)
Comment (simanadler) [reply manager/controllers/app/blueprint_controller.go:92]: Will change

File: manager/controllers/app/blueprint_controller.go (hunk 5 above)
Comment (simanadler) [reply manager/controllers/app/blueprint_controller.go:154]: OK, changing

File: manager/controllers/app/blueprint_controller.go (hunk 6 above)
Comment (simanadler) [reply manager/controllers/app/blueprint_controller.go:286]: OK, changing

File: manager/controllers/app/blueprint_controller.go (hunk 7 above)
Comment (simanadler) [reply manager/controllers/app/blueprint_controller.go:311]: FYBRIKAPPUID will be used in both the control plane and in the data plane, so I didn't think it belonged in app.  Regarding the capital letters, I will change it

File: manager/controllers/app/fybrikapplication_controller.go (hunk 9 above)
Comment (simanadler) [reply manager/controllers/app/fybrikapplication_controller.go:103]: OK

File: manager/controllers/app/fybrikapplication_controller.go (hunk 10 above)
Comment (simanadler) [reply manager/controllers/app/fybrikapplication_controller.go:205]: Will change

File: manager/controllers/app/fybrikapplication_controller.go (hunk 11 above)
Comment (simanadler) [reply manager/controllers/app/fybrikapplication_controller.go:357]: OK

File: manager/controllers/app/select_modules.go [hunk 12]
```diff
@@ -126,7 +127,7 @@ func (p *PlotterGenerator) validate(item *DataInfo, solution Solution, appContex
 				element.StorageAccountRegion = region
 			}
 			if element.StorageAccountRegion == "" {
-				p.Log.V(0).Info("Could not find a storage account, aborting data path construction")
+				p.Log.Fatal().Str(logging.DATASETID, item.Context.DataSetID).Msg("Could not find a storage account, aborting data path construction")
```
Comment (shlomitk1) [inline manager/controllers/app/select_modules.go:130]: Debug, not Fatal. It's a step where some of the data paths are invalidated. If copy can not be done, and is not required either, it is not a mistake to skip a data path with copy.

File: manager/controllers/app/select_modules.go [hunk 13]
```diff
@@ -143,13 +144,13 @@ func (p *PlotterGenerator) validate(item *DataInfo, solution Solution, appContex
 		requiredActions = unsupported
 		// select a cluster for the capability that satisfy cluster restrictions specified in admin config policies
 		if !p.findCluster(item, element, appContext) {
-			p.Log.V(0).Info("Could not find an available cluster for " + moduleCapability.Capability)
+			p.Log.Error().Str(logging.DATASETID, item.Context.DataSetID).Msg("Could not find an available cluster for " + moduleCapability.Capability)
```
Comment (shlomitk1) [inline manager/controllers/app/select_modules.go:147]: Debug, not Error. A path is invalidated if a capability along the path can not be deployed.

File: manager/controllers/app/select_modules.go [hunk 14]
```diff
@@ -143,13 +144,13 @@ func (p *PlotterGenerator) validate(item *DataInfo, solution Solution, appContex
...
 			return false
 		}
 	}
 	// Are all actions supported by the capabilities in this data path?
 	if len(requiredActions) > 0 {
-		p.Log.V(0).Info("Not all governance actions are supported, aborting data path construction")
+		p.Log.Fatal().Str(logging.DATASETID, item.Context.DataSetID).Msg("Not all governance actions are supported, aborting data path construction")
```
Comment (shlomitk1) [inline manager/controllers/app/select_modules.go:153]: Debug, not Fatal. Path is invalidated if not all actions can be performed along the path.

File: manager/controllers/app/select_modules.go (hunk 12 above)
Comment (simanadler) [reply manager/controllers/app/select_modules.go:130]: OK

File: manager/apis/app/v1alpha1/zz_generated.deepcopy.go [hunk 15]
```diff
@@ -1,5 +1,3 @@
-// +build !ignore_autogenerated
```
Comment (shlomitk1) [inline manager/apis/app/v1alpha1/zz_generated.deepcopy.go:1]: Please restore this file.

File: site/docs/contribute/logging.md [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+
+# Log Entry Verbosity
+Log levels should be used as follows:
+
+- panic (zerolog.PanicLevel, 5) - Errors that prevent the component from operating correctly and handling requests
+     Ex: fybrik control plane did not deploy correctly
+	   Ex: Data plane component crashed and cannot handle requests
+- fatal (zerolog.FatalLevel, 4) - Errors that prevent the component from successfully completing a particular task
+	   Ex: fybrikapplication controller cannot generate a plotter
+	   Ex: Arrow/Flight server used to read data cannot access data store
```
Comment (shlomitk1) [inline site/docs/contribute/logging.md:27]: I don't think `fatal` should be used at all. The examples above report errors from a single application without causing any disturbance to other users. We don't want the pod to be restarted in this case.

File: manager/controllers/utils/utils.go [hunk 17]
```diff
@@ -157,3 +157,22 @@ func Intersection(set1 []string, set2 []string) []string {
 	}
 	return res
 }
+
+const FybrikAppUUID = "FybrikApplicationUUID"
```
Comment (shlomitk1) [inline manager/controllers/utils/utils.go:161]: I think the convention is to use something like `fybrik-application-uuid`

**Review by shlomitk1 — APPROVED** [review]
